### PR TITLE
Test pull request for peer review.

### PR DESCRIPTION
**Pull request by SilverSynch**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,32 +1,136 @@
...
+        if (wordListCache == null) {
+            throw new NullPointerException("No word list cache is available or string was not set!");
+        }
+        return wordListCache.length;
     }
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        if (internalString == null) {
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:50]: I didn't include a check for if my string is null so good job implementing this even though it wasn't specified!  :)

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,32 +1,136 @@
...
+            throw new NullPointerException("String was not set!");
+        }
+        else if (n > internalString.length()) {
+            throw new IndexOutOfBoundsException("Removal count is greater than the length of the string (%s).".formatted(internalString.length()));
+        }
+        else if (n < 1) {
+            throw new IllegalArgumentException("Removal count is less than 1.");
+        }
+
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:59]: See my version for a simpler version of the code!

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,32 +1,136 @@
...
+    // In a more professional enviroment, maintaining cache validation would be mandatory.
+    private String[] wordListCache;
+
+    /**
+     * Generates or regenerated the internal cache of words in the string.
+     * You must call this method every time the internal string is modified!
+     */
+    private void genWordListCache() {
+        // Splits the string by words.
+        // Matches any cluster of non-whitespace characters.
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:17]: What I did here is first I trimmed leading and lagging whitespace using String.trim().
Then I also used String.split() but provided the regex " +" for 1 or more space characters. 

This regex can be changed to "\\s+" for 1 or more whitespace characters. 

I found that by using the trim() method my issue with the regex matching an empty string was solved.

File: src/main/java/StringManipulation.java (hunk 2 above)
Comment (NadavH12) [reply src/main/java/StringManipulation.java:59]: My (maintainSpacing) is pretty much the same as yours. 

What I did for the (!maintainSpacing) case is that I just replaced the n'th char with an "empty char".
The for loop variable was still tracking the string index properly. 

I'm not sure if that makes sense but let me know if you have any questions and I am happy to answer them, or feel free to look at my code.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,32 +1,136 @@
...
     public void setString(String string) {
+        internalString = string;
+        genWordListCache();
     }
 
     @Override
     public int count() {
-        return 0;
+        if (wordListCache == null) {
+            throw new NullPointerException("No word list cache is available or string was not set!");
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:43]: I did a similar thing here with using the String.split() method and then returning the length of that array for count.

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,32 +1,136 @@
...
+            throw new IllegalArgumentException("endWord cannot be less than 1.");
+            
+        }
+        else if (endWord < count()) {
+            throw new IndexOutOfBoundsException("endWord is greater than the number of words in the string (%s).".formatted(count()));
+        }
+
+        String[] result = new String[endWord - startWord + 1];
+
+        for (int i = 0; i <= endWord - startWord; i++) {
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:105]: Equivalent Code:
result = Arrays.copyOfRange(wordsListCache, startword - 1, endword);

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,32 +1,136 @@
...
+            result[i] = wordListCache[startWord - 1 + i];
+        }
+        
+        return result;
     }
 
     @Override
-    public String restoreString(int[] indices) {
-        return null;
+    public String restoreString(int[] indices){
```
Comment (NadavH12) [inline src/main/java/StringManipulation.java:113]: 😄

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -26,19 +26,26 @@ public void testCount1() {
         assertEquals(4, length);
     }
 
+    // Test that unset strings throw NullPointerException.
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:29]: 👍

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -26,19 +26,26 @@ public void testCount1() {
...
     @Test
     public void testCount2() {
-        fail("Not yet implemented");
+        assertThrows(NullPointerException.class, () -> {manipulatedstring.count();});
     }
 
+    // Test that empty strings have a word count of 0.
     @Test
     public void testCount3() {
-        fail("Not yet implemented");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:36]: 👍

File: src/test/java/StringManipulationTest.java [hunk 9]
```diff
@@ -26,19 +26,26 @@ public void testCount1() {
...
     }
 
+    // Test that a string of 1 single word but a trailing space has a word count of 1.
     @Test
     public void testCount4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("superfragilisticalidocious ");
+        int length = manipulatedstring.count();
+        assertEquals(1, length);
     }
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:50]: Include test case for multi word strings, and a test case for strings with weird whitespace.

File: src/test/java/StringManipulationTest.java [hunk 10]
```diff
@@ -53,29 +60,38 @@ public void testRemoveNthCharacter2() {
         assertEquals("I'  b tt r  ut s0 e  16 ts in th s  tr n6  r gh ?", manipulatedstring.removeNthCharacter(3, true));
     }
 
+    // Test that removal counts greater than the length of the string throw IndexOutOfBoundsException.
     @Test
     public void testRemoveNthCharacter3() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("only");
+        assertThrows(IndexOutOfBoundsException.class, () -> {manipulatedstring.removeNthCharacter(5, false);});
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:67]: 👍

File: src/test/java/StringManipulationTest.java [hunk 11]
```diff
@@ -53,29 +60,38 @@ public void testRemoveNthCharacter2() {
...
     }
 
+    // Test that removal counts smaller than 1 throw IllegalArgumentException.
     @Test
     public void testRemoveNthCharacter4() {
-        fail("Not yet implemented");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:63]: 👍

File: src/test/java/StringManipulationTest.java [hunk 12]
```diff
@@ -53,29 +60,38 @@ public void testRemoveNthCharacter2() {
...
     @Test
     public void testRemoveNthCharacter5() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("    e\t");
+        assertEquals("   e", manipulatedstring.removeNthCharacter(3, false));
     }
 
+    // Test that removal counts of 0 on strings with lengths of 0 still throw IllegalArgumentException.
     @Test
     public void testRemoveNthCharacter6() {
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:86]: 👍

File: src/test/java/StringManipulationTest.java [hunk 13]
```diff
@@ -53,29 +60,38 @@ public void testRemoveNthCharacter2() {
...
-        fail("Not yet implemented");
+        manipulatedstring.setString("Haha");
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.removeNthCharacter(0, false);});
     }
 
+    // Test that removal removes all characters, including whitespace characters.
     @Test
     public void testRemoveNthCharacter5() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("    e\t");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:80]: 👍

File: src/test/java/StringManipulationTest.java [hunk 14]
```diff
@@ -53,29 +60,38 @@ public void testRemoveNthCharacter2() {
...
-        fail("Not yet implemented");
+        manipulatedstring.setString("");
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.removeNthCharacter(0, false);});
     }
 
+    // Test that removal on unset strings throw NullPointerException.
     @Test
     public void testRemoveNthCharacter7() {
-        fail("Not yet implemented");
+        assertThrows(NullPointerException.class, () -> {manipulatedstring.removeNthCharacter(2, false);});
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:94]: 👍

File: src/test/java/StringManipulationTest.java [hunk 15]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
         assertEquals(sStings[1], "string");
     }
 
+    // Test that endWord indices greater than the length of the string throw IndexOutOfBoundsException.
     @Test
     public void testGeSubStrings2() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("You");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:109]: 👍

File: src/test/java/StringManipulationTest.java [hunk 16]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
...
+        assertThrows(IndexOutOfBoundsException.class, () -> {manipulatedstring.getSubStrings(1, 5);});
     }
+
+    // Test that startWord indices less than 1 throw IllegalArgumentException.
     @Test
     public void testGeSubStrings3() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("You");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:116]: 👍

File: src/test/java/StringManipulationTest.java [hunk 17]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
...
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.getSubStrings(0, 3);});
     }
+
+    // Test that endWord indices less than 1 throw IllegalArgumentException.
     @Test
     public void testGeSubStrings4() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("You");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:123]: 👍

File: src/test/java/StringManipulationTest.java [hunk 18]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
...
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.getSubStrings(-3, -1);});
     }
+
+    // Test that endWord indices less than the startWord indices throw IllegalArgumentException.
     @Test
     public void testGeSubStrings5() {
-        fail("Not yet implemented");
+        manipulatedstring.setString("You");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:130]: 👍

File: src/test/java/StringManipulationTest.java [hunk 19]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
...
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.getSubStrings(2, 1);});
     }
+
+    // Test that substring grabs on unset strings throw NullPointerException.
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:134]: 👍

File: src/test/java/StringManipulationTest.java [hunk 20]
```diff
@@ -87,25 +103,38 @@ public void testGeSubStrings1() {
...
     @Test
     public void testGeSubStrings6() {
-        fail("Not yet implemented");
+        assertThrows(NullPointerException.class, () -> {manipulatedstring.getSubStrings(1, 5);});
     }
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:139]: Include test cases for valid inputs with funky white-spacing.

File: src/test/java/StringManipulationTest.java [hunk 21]
```diff
@@ -118,32 +147,39 @@ public void testRestoreString1()
         assertEquals(restoreString, "rat");
     }
 
+    // Test that restore arrays less than the length of the string throw IllegalArgumentException.
     @Test
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:151]: 👍

File: src/test/java/StringManipulationTest.java [hunk 22]
```diff
@@ -118,32 +147,39 @@ public void testRestoreString1()
...
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("That");
+        int[] array = {1, 0, 2};
+        assertThrows(IllegalArgumentException.class, () -> {manipulatedstring.restoreString(array);});
     }
 
+    // Test that restore arrays with indices less than 0 throw IndexOutOfBoundsException.
     @Test
     public void testRestoreString3()
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:161]: 👍

File: src/test/java/StringManipulationTest.java [hunk 23]
```diff
@@ -118,32 +147,39 @@ public void testRestoreString1()
...
+        assertThrows(IndexOutOfBoundsException.class, () -> {manipulatedstring.restoreString(array);});
     }
 
+    // Test that arrays with indices greater than or equal to the length of the string throw IndexOutOfBoundsException.
     @Test
     public void testRestoreString4()
     {
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("That");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:172]: 👍

File: src/test/java/StringManipulationTest.java [hunk 24]
```diff
@@ -118,32 +147,39 @@ public void testRestoreString1()
...
+        int[] array = {1, 0, 2, 4};
+        assertThrows(IndexOutOfBoundsException.class, () -> {manipulatedstring.restoreString(array);});
     }
 
+    // Test that restoration on unset strings throw NullPointerException.
     @Test
     public void testRestoreString5()
     {
-        fail("Not yet implemented");
```
Comment (NadavH12) [inline src/test/java/StringManipulationTest.java:145]: 👍

File: README.md
Comment (NadavH12) [inline README.md:1]: Overall nice code. Code really shows that you have a strong grasp on the higher level features of Java. My advice to you is to try to keep things as simple as possible. Although a roundabout way might work I think that the people who developed these exercises have a simple solution in mind and it is our job to try and find that simple solution. Also be more creative with your test cases, always throw in some random whitespace into inputs because it's guaranteed to break shit. 👍 😄

**Review by NadavH12 — APPROVED** [review]
